Diagnostic/diagnostic.py

```diff
@@ -828,6 +848,21 @@ def start_mdsd():
 
                 # mdsd is now up for at least 30 seconds.
 
+                # Mitigate if memory leak is suspected.
+                mdsd_memory_leak_suspected, mdsd_memory_usage_in_MB = check_suspected_memory_leak(mdsd.pid)
+                if mdsd_memory_leak_suspected:
+                    memory_leak_msg = "Suspected mdsd memory leak (memory usage: {0}MB). Recycling mdsd to self-mitigate.".format(mdsd_memory_usage_in_MB)
+                    hutil.log(memory_leak_msg)
+                    # Add a telemetry as well. This might ding our enable success KPI,
+                    # but there doesn't seem to be any other way.
+                    waagent.AddExtensionEvent(name=hutil.get_name(),
+                                              op=ExtensionOperationType,
+                                              isSuccess=False,
+                                              version=hutil.get_extension_version(),
+                                              message=memory_leak_msg)
+                    mdsd.kill()
```

Comment:
In the next iteration of the outer loop. This particular python function is far too large to be easily understood, and Python's syntax doesn't help. I had the same question and it took me five minutes of scrolling to find exactly what happens after the ```break``` statement on 864, but sure enough eventually it comes right back to the top of the function (near enough).